src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs

```diff
@@ -0,0 +1,212 @@
+
+            foreach (QueueWithEvents queueItem in items)
+            {
+                var topicName = queueItem.Name;
+
+                foreach (var insertItem in queueItem.Inserts)
+                {
+                    var eventMessage = new CloudEvent
+                    {
+                        Id = DefaultCloudEventId,
+                        DataContentType = "application/json",
+                        Type = string.IsNullOrWhiteSpace(queueItem.Type) ? DefaultType : queueItem.Type,
+                        Source = new Uri(string.IsNullOrWhiteSpace(queueItem.Source) ? DefaultSource : queueItem.Source),
+                        Data = insertItem.TransformationResult
+                    };
+
+                    eventMessage.SetPartitionKey(insertItem.ChangeVector);
+
+                    events.Add(new KafkaMessageEvent
+                    {
+                        Topic = topicName, Message = eventMessage.ToKafkaMessage(ContentMode.Binary, jsonSerializer)
+                    });
+                }
+            }
+
+            if (events.Count == 0) return count;
+
+            if (_kafkaProducer == null)
+            {
+                _kafkaProducer = QueueHelper.CreateKafkaClient(Configuration.Connection, TransactionalId,
+                    Database.ServerStore.Server.Certificate.Certificate);
+                _kafkaProducer.InitTransactions(TimeSpan.FromSeconds(60));
+            }
+
+            void ReportHandler(DeliveryReport<string?, byte[]> report)
+            {
+                if (report.Error.IsError == false)
+                {
+                    count++;
+                }
+                else
+                {
+                    Logger.Info($"Failed to deliver message: {report.Error.Reason}");
+                    _kafkaProducer.AbortTransaction();
+                }
+            }
+
+            try
+            {
+                _kafkaProducer.BeginTransaction();
+
+                foreach (var @event in events)
+                {
+                    _kafkaProducer.Produce(@event.Topic, @event.Message, ReportHandler);
+                }
+
+                _kafkaProducer.CommitTransaction();
+            }
+            catch (Exception ex)
+            {
+                if (ex is KafkaException) _kafkaProducer.AbortTransaction();
+                throw new QueueLoadException(ex.Message, ex);
+            }
+        }
+
+        return count;
+    }
+
+    private void ProcessRabbitMq()
+    {
+    }
+
+    public QueueEtlTestScriptResult RunTest(IEnumerable<QueueWithEvents> records, DocumentsOperationContext context)
+    {
+        var simulatedWriter = new QueueWriterSimulator();
+        var summaries = new List<TopicSummary>();
+
+        foreach (var record in records)
+        {
+            var commands = simulatedWriter.SimulateExecuteCommandText(record, context);
+
+            summaries.Add(new TopicSummary { TopicName = record.Name, Commands = commands.ToArray() });
```

Comment:
In here, should we also send to Kafka and then abort? 
IIRC, we do something similar for SQL ETL - to ensure things like schema validation

Reply:
We don't do that in Kafka, because consumers configured with read uncommitted option can consume the message before we abort the transaction.